### ISSUE TITLE
To run nightly builds of Bibledit for all architectures

### ISSUE THREAD

**Issue by teusbenschop** [bibledit/bibledit#125]

**Comment by alerque** [comment]
The repo and makefile cleanup I've been whittling away at should make the build process a lot more predictable.  With the process cleaned up, we should be able to pass all commits through a CI server to see whether they break the builds or not. The CI server could pass the build artifacts on to an archive of builds, so "nightly" wouldn't be needed, there would just be downloads available for successful builds.

**Comment by teusbenschop** [comment]
This continuous integration server would be helpful.
And what would be helpful too would be if the CI server would have some automated GUI testing - it's not being done now.

**Comment by teusbenschop** [comment]
Will the continuous integration server that you propose also be able to build Mac and iOS and Android and Chrome OS and other such exotic builds, or will it just build for Linux and Windows?

**Comment by alerque** [comment]
Travis has OSX instances available, so I expect the Mac and iOS builds are serviceable from there. Android can likewise be build from Linux, as can ChromeOS I think. Windows is actually the one I don't know enough about how the build chain works to say. If it can be cross compiled from a Linux host I'm sure we can do it, but Travis doesn't have Windows build servers. There are probably such a thing out there, I just don't know what the CI build scene looks like for Windows software.

**Comment by teusbenschop** [comment]
It is helpful that Travis has OSX instances available as well.
As for Windows, BE is now built from a Cygwin terminal. The Windows GUI wrapper uses .NET. Whether the whole app for Windows can be cross-compiled, I don't know - perhaps it can, when enough effort is put in.
If MS Windows would die a death, that would help, we would no longer have to worry about it.

**Comment by teusbenschop** [comment]
I thought to work on this one today, but it would be better to postpone this till you have done the cleanup of the repository and the Makefile-s. 

Even so I am writing a a few notes for a possible implementation.

Currently supported operating systems:
WINDOWS
ANDROID
MAC
LINUX
IOS
CHROME OS
CLOUD

If the build logs are public on Travis, that would be great.

Else it would also help to create a special area on bibledit.org where the script output appears after every build.
Each time to upload only the build logs, not the whole of the website, and to store the build logs also on the local copy of the website, else publishing the website would delete the build logs.

It's good to create a file for each build type, and to name the files very clearly.

An option would be to link to the build files from the bibledit demo menu help / installation, section development.

[end notes] :)

**Comment by alerque** [comment]
Sorry the repository migration thing isn't finished yet. I don't want to jump the gun because once it's done it's hard to go back and change without being disruptive. There has been progress, but other things keep taking my time. This week it's sermon prep ;-)

Anyway do know that I have ants in my pants to finish this up for my own reasons of needing to get some other stuff deployed for our team that is kind of held up by the BE upgrade process being so hard.

One thing that's starting to be apparent through my hacking on the repos and your comments here is that we have two related but separate needs from the CI system:
1. Running a test suite on the `bibledit-core` repository
2. Building platform specific releases from the `bibledit-<platform>` repositories

This brings up the question of how we want the release cycle to work. Obviously the  CI server should check what's up on every commit to `bibledit-core`, but should this also trigger builds for every platform? For testing purposes this seems like it should be a yes, but how will the _release_ cycles actually  work? Will they only update when there are platform specific changes? Periodically to refresh the core stuff? What should the version number scheme look like? Should the platform version be the major component of the semantic version number string or should the core version dictate that with the platform version being a sub-part?

Answers to these questions will dictate how the submodules are linked, how versions numbers are bumped, and what build jobs are fired off when.

(For reference, issues #171 and #172 are related.)

**Comment by teusbenschop** [comment]
No need to worry about the progress on this migration thing to another repository, it can be done whenever there's time, no hurry at all.

But I wonder what part of the Bibledit upgrade process is so hard from your perspective, as I don't know of anything that is hard from my perspective, as all looks easy to do. If there's something considered hard to do, then I may look at how to make that easier.

As for the release cycle, some projects do code-freezes, to give enough time to a release to become stable enough to be released. Though we don't have a code-free in place right now, I do always take a hard look at the state of things to know whether to release for a certain platform, so that we know that something is released that is working. It means that we cannot automatically release things.

It would be good to have a fast release cycle, I think, because the code right now is being changed fast. Each release of course need to have the best and most recent core code in them. A monthly release cycle sounds good, with interim release when something important gets fixed that affects the users in the field.

I am not particularly concerned about the version numbering scheme, and am happy with the daily bumping of the version number, though I can imagine there are better ways of doing it - you may decide what to do. What is important for most operating systems is that they can deduce which version is newer, for the upgrade system to work properly.

**Comment by teusbenschop** [comment]
The daily build logs are now available from http://bibledit.org/builds.
